spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/ZookeeperMetadataStore.java

```diff
@@ -0,0 +1,381 @@
+	private final ConcurrentMap<String, LocalChildData> updateMap = new ConcurrentHashMap<>();
+
+	private volatile String root = "/SpringIntegration-MetadataStore";
+
+	private volatile String encoding = "UTF-8";
+
+	private volatile PathChildrenCache cache;
+
+	private volatile boolean running = false;
+
+	private volatile boolean autoStartup = false;
+
+	private volatile int phase = Integer.MAX_VALUE;
+
+	public ZookeeperMetadataStore(CuratorFramework client) throws Exception {
+		Assert.notNull(client, "Client cannot be null");
+		this.client = client;
+	}
+
+	/**
+	 * Encoding to use when storing data in ZooKeeper
+	 *
+	 * @param encoding encoding as text
+	 */
+	public void setEncoding(String encoding) {
+		Assert.notNull(encoding, "'encoding' must not be null.");
+		this.encoding = encoding;
+	}
+
+	/**
+	 * Root node - store entries are children of this node
+	 *
+	 * @param root encoding as text
+	 */
+	public void setRoot(String root) {
+		Assert.notNull(root, "'root' must not be null.");
+		Assert.isTrue(root.startsWith("/"), "'root' must start with '/'");
+		Assert.isTrue(!root.endsWith("/"), "'root' must not end with '/'");
+		this.root = root;
+	}
+
+	public String getRoot() {
+		return this.root;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public String putIfAbsent(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		synchronized (this.updateMap) {
+			try {
+				createNode(key, value);
+				return null;
+			}
+			catch (KeeperException.NodeExistsException e) {
+				// so the data actually exists, we can read it
+				try {
+					byte[] bytes = this.client.getData().forPath(getPath(key));
+					return Conversions.bytesToString(bytes, "UTF-8");
+				}
+				catch (Exception exceptionDuringGet) {
+					throw new ZookeeperMetadataStoreException("Exception while creating node with key '" + key + "':", e);
+				}
+			}
+			catch (Exception e) {
+				throw new ZookeeperMetadataStoreException("Error while trying to set '" + key + "':", e);
+			}
+		}
+	}
+
+	@Override
+	public boolean replace(String key, String oldValue, String newValue) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(oldValue, "'oldValue' must not be null.");
+		Assert.notNull(newValue, "'newValue' must not be null.");
+		synchronized (this.updateMap) {
+			Stat currentStat = new Stat();
+			try {
+				byte[] bytes = this.client.getData().storingStatIn(currentStat).forPath(getPath(key));
+				if (oldValue.equals(Conversions.bytesToString(bytes, "UTF-8"))) {
+					updateNode(key, newValue, currentStat.getVersion());
+				}
+				return true;
+			}
+			catch (KeeperException.NoNodeException e) {
+				// ignore, the node doesn't exist there's nothing to replace
+				return false;
+			}
+			catch (KeeperException.BadVersionException e) {
```

Comment:
Correct. In fact, optimistic locking is much better in this case, it guarantees correct behaviour even if changes are made directly in ZK.